src/Storages/HDFS/StorageHDFSCluster.cpp

```diff
@@ -140,6 +134,29 @@ QueryProcessingStage::Enum StorageHDFSCluster::getQueryProcessingStage(
 }
 
 
+void StorageHDFSCluster::createIteratorAndCallback(ContextPtr context) const
```

Comment:
Why can't this function return cluster and callback? It's a bit confusing to have it modify member values while calling it always before using them.
Don't see a lifetime issue as the `shared_ptr` are copied/used instantly.

Or even better, 2 functions for iterator and cluster.

Reply:
I was a mess when I wrote this. Generally I thought this "initialization" could happen twice and thus made such an API